feat: add C `ndarray` interface and refactor implementation for `stats/base/dsnanmean`

Neerajpathak07 opened the pull request:
Resolves none.

## Description

> What is the purpose of this pull request?

This pull request:

-  adds C `ndarray` interface for `stats/base/dsnanmean`
-   refactor JavaScript implementation
-   refactor examples, tests, docs and benchmarks.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves none

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: ndarray-dsnanmean

Neerajpathak07 committed:
fix: lint error

Neerajpathak07 committed:
fix: CI error

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/README.md:
```diff
@@ -181,6 +181,107 @@ console.log( v );
+
+#### stdlib_strided_dsnanmean( N, \*X, strideX )
+
+Calculate the Arithmetic Mean value of a double-precision floating-point strided array, ignoring `NaN` values.
```
This description should more closely match that of the JS APIs above.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/README.md:
```diff
@@ -181,6 +181,107 @@ console.log( v );
+
+#### stdlib_strided_dsnanmean_ndarray( N, \*X, strideX, offsetX )
+
+Computes the Arithmetic Mean value of a double-precision floating-point strided array, ignoring `NaN` values and using alternative indexing semantics.
```
Same comment. I am also not sure why "arithmetic mean" is capitalized.

kgryte replied:
The description is also incorrect. `X` is not double-precision, but single-precision.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/README.md:
```diff
@@ -181,6 +181,107 @@ console.log( v );
+
+int main( void ) {
+    // Create a strided array:
+    const float x[] = { 1.0f, 2.0f, 0.0f/0.0f, 3.0f, 0.0f/0.0f, 4.0f, 5.0f, 6.0f,   0.0f/0.0f, 7.0f, 8.0f, 0.0f/0.0f };
```
```suggestion
    const float x[] = { 1.0f, 2.0f, 0.0f/0.0f, 3.0f, 0.0f/0.0f, 4.0f, 5.0f, 6.0f, 0.0f/0.0f, 7.0f, 8.0f, 0.0f/0.0f };
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/benchmark.js:
```diff
@@ -39,17 +45,7 @@ var dsnanmean = require( './../lib/dsnanmean.js' );
-			x[ i ] = ( randu()*20.0 ) - 10.0;
-		}
-	}
+	var x = uniform( len, -10.0, 10.0, options );
```
This doesn't match the original random value generation. Need to use `filledarrayBy` and a custom random function.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/benchmark.ndarray.js:
```diff
@@ -39,17 +45,7 @@ var dsnanmean = require( './../lib/ndarray.js' );
-			x[ i ] = ( randu()*20.0 ) - 10.0;
-		}
-	}
+	var x = uniform( len, -10.0, 10.0, options );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/benchmark.ndarray.native.js:
```diff
@@ -48,17 +50,7 @@ var opts = {
-			x[ i ] = ( randu()*20.0 ) - 10.0;
-		}
-	}
+	var x = uniform( len, -10.0, 10.0, options );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -94,19 +94,20 @@ static float rand_float( void ) {
 	double elapsed;
 	float x[ len ];
-	double v;
+	float v;
```
This is incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -94,19 +94,20 @@ static float rand_float( void ) {
 		x[ i ] = ( rand_float()*20000.0f ) - 10000.0f;
 	}
-	v = 0.0;
+	v = 0.0f;
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+static double benchmark2( int iterations, int len ) {
+	double elapsed;
+	float x[ len ];
+	float v;
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = ( rand_float()*20000.0f ) - 10000.0f;
+	}
+	v = 0.0f;
```
```suggestion
	v = 0.0;
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/docs/repl.txt:
```diff
@@ -22,8 +22,8 @@
-    stride: integer
-        Index increment.
+    strideX: integer
+        Stride Length.
```
```suggestion
        Stride length.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/docs/repl.txt:
```diff
@@ -39,26 +39,23 @@
-    > {{alias}}( N, x, stride )
-    ~0.3333
+    > {{alias}}( 6, x, 2 )
+    NaN
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/docs/repl.txt:
```diff
@@ -39,26 +39,23 @@
-    > stride = 2;
-    > {{alias}}( N, x1, stride )
-    ~-0.3333
+    > {{alias}}( 6, x1,2 )
+    NaN
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/docs/repl.txt:
```diff
@@ -69,10 +66,10 @@
-    stride: integer
-        Index increment.
+    strideX: integer
+        Stride Length.
```
```suggestion
        Stride length.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/docs/repl.txt:
```diff
@@ -89,9 +86,8 @@
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > {{alias}}.ndarray( N, x, 2, 1 )
-    ~-0.3333
+    > {{alias}}.ndarray( 6, x, 2, 1 )
+    NaN
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/examples/index.js:
```diff
@@ -18,22 +18,12 @@
-		x[ i ] = round( (randu()*100.0) - 50.0 );
-	}
-}
+var x = discreteUniform( 10, -50, 50, {
```
This doesn't match the original random value generation logic.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/include/stdlib/stats/base/dsnanmean.h:
```diff
@@ -31,7 +31,12 @@ extern "C" {
+double API_SUFFIX(stdlib_strided_dsnanmean)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX );
+
+/**
+* Computes the mean absolute value of a double-precision floating-point strided array, ignoring `NaN` values and using alternative indexing semantics .
```
Doesn't agree with description above.

kgryte replied:
The description is also wrong.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/lib/index.js:
```diff
@@ -28,20 +28,17 @@
-* var N = floor( x.length / 2 );
 *
-* var v = dsnanmean.ndarray( N, x, 2, 1 );
+* var v = dsnanmean.ndarray( 5, x, 2, 1 );
```
This isn't correct. What was the computed value of `N` previously?

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/lib/ndarray.js:
```diff
@@ -30,22 +30,17 @@ var dsnanmeanpn = require( '@stdlib/stats/base/dsnanmeanpn' ).ndarray;
 * var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN ] );
-* var N = floor( x.length / 2 );
-*
-* var v = dsnanmean( N, x, 2, 1 );
-* // returns 1.25
```
Why was this removed? Don't remove inconvenient code (e.g., if lint errors are happening, fix the lint errors; don't just delete it).

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/lib/ndarray.native.js:
```diff
@@ -31,27 +30,17 @@ var addon = require( './dsnanmean.native.js' );
 * var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN ] );
-* var N = floor( x.length / 2 );
-*
-* var v = dsnanmean( N, x, 2, 1 );
-* // returns 1.25
```
Again. Why was this deleted?

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/manifest.json:
```diff
@@ -24,17 +27,83 @@
+      "dependencies": [
+        "@stdlib/blas/base/shared",
+        "@stdlib/strided/base/stride2offset",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/assert/is-positive-zero",
```
TMK, neither of these dependencies are used. Please don't just blindly copy-paste things from other packages. Think about what you are doing and whether something is applicable to the specific logic/implementation at hand.

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/src/main.c:
```diff
@@ -18,16 +18,31 @@
+}
+
+/**
+* Computes the maximum value of a double-precision floating-point strided array, ignoring `NaN` values and using alternative indexing semantics.
```
How does this description make sense?

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/test/test.ndarray.js:
```diff
@@ -106,32 +104,7 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
-	t.end();
-});
-
-tape( 'the function supports a negative `stride` parameter', function test( t ) {
```
Why was this test removed?

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/test/test.ndarray.native.js:
```diff
@@ -115,32 +113,7 @@ tape( 'the function supports a `stride` parameter', opts, function test( t ) {
-	t.end();
-});
-
-tape( 'the function supports a negative `stride` parameter', opts, function test( t ) {
```
Why was this test removed?

kgryte reviewed (changes requested):
Left comments. Please do a very thorough review of all your open PRs implementing similar changes. They are currently inundated with copy-paste mistakes.

Neerajpathak07 commented:
@kgryte Thank you for you review. Extremely sorry for the errors. I'll fix these mistakes for other PR's as well.

Neerajpathak07 committed:
chore: minor clean up

Neerajpathak07 committed:
fix: lint copyright years

Neerajpathak07 commented:
@kgryte 
I have made all the essential changes to all the packages requiring similar implementations.
This time I was able to understand what the package was demanding for and how to go about it.

Neerajpathak07 committed:
chore: bench clean up

Neerajpathak07 committed:
Merge branch 'stdlib-js:develop' into ndarray-dsnanmean

Neerajpathak07 committed:
chore: update grammar

Neerajpathak07 committed:
fix: lint copyright years

Neerajpathak07 committed:
Merge branch 'ndarray-dsnanmean' of https://github.com/Neerajpathak07/stdlib into ndarray-dsnanmean

Neerajpathak07 committed:
chore: markdown changes

Neerajpathak07 commented:
@kgryte I'm awaiting the PR with `dsnanmeanpn` to be merged then I'll be able to reference this implementation.

Neerajpathak07 commented:
@kgryte I have made the suggested changes throughout all my other PR's would really appreciate a review

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dsnanmean/src/main.c:
```diff
@@ -0,0 +1,48 @@
+* @return         output value
+*/
+double API_SUFFIX(stdlib_strided_dsnanmean_ndarray)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	return stdlib_strided_dsnanmeanpn_ndarray( N, X, strideX, offsetX );
```
@kgryte I'm awaiting the PR with `dsnanmeanpn` to be merged then I'll be able to reference this implementation.

Neerajpathak07 replied:
PR with `dsnanmeanpn` has been merged! #4405

Neerajpathak07 committed:
chore: minor clean up

Neerajpathak07 committed:
chore: update addon

Neerajpathak07 committed:
Merge branch 'develop' into ndarray-dsnanmean

Signed-off-by: Neeraj Pathak <neerajrpathak710@gmail.com>

Neerajpathak07 committed:
chore: clean up

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

Neerajpathak07 committed:
chore: max-len

Neerajpathak07 committed:
chore: max-len

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dsnanmean/README.md:
```diff
@@ -181,6 +177,107 @@ console.log( v );
 
 <!-- /.examples -->
 
```
readme structure is incorrect

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dsnanmean/docs/repl.txt:
```diff
@@ -37,28 +37,25 @@
-    > N = {{alias:@stdlib/math/base/special/floor}}( x0.length / 2 );
-    > stride = 2;
-    > {{alias}}( N, x1, stride )
+    > {{alias}}( 4, x1,2 )
```
spacing
```suggestion
    > {{alias}}( 4, x1, 2 )
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dsnanmean/include/stdlib/stats/base/dsnanmean.h:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2020 The Stdlib Authors.
+* Copyright (c) 2025 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2020 The Stdlib Authors.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dsnanmean/src/main.c:
```diff
@@ -0,0 +1,48 @@
+* @return         output value
+*/
+double API_SUFFIX(stdlib_strided_dsnanmean_ndarray)( const CBLAS_INT N, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	return stdlib_strided_dsnanmeanpn_ndarray( N, X, strideX, offsetX );
```
wrap in `API_SUFFIX`

aayush0325 committed:
chore: code review

aayush0325 reviewed (dismissed):
this should be clear of any errors now

Neerajpathak07 commented:
@kgryte could we land this in.

Neerajpathak07 commented:
@kgryte pls take a look at this as well!!

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/README.md:
```diff
@@ -181,6 +177,123 @@ console.log( v );
+    double v = stdlib_strided_dsnanmean( N, x, strideX );
+
+    // Print the result:
+    printf( "mean: %f\n", v );
```
```suggestion
    printf( "mean: %lf\n", v );
```

`v` is a double.

kgryte committed:
docs: fix example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dsnanmean/manifest.json:
```diff
@@ -46,33 +50,52 @@
+      ]
+    },
+    {
+      "task": "",
```
```suggestion
      "task": "build",
```

kgryte committed:
chore: update build task

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
